### PR TITLE
ed25519 implementation

### PR DESCRIPTION
**Pull request by DieMyst**

### CODE REVIEW

File: hashsign/jvm/src/main/scala/fluence/crypto/ecdsa/Ed25519.scala [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+ * Edwards-curve Digital Signature Algorithm (EdDSA)
+ */
+class Ed25519(strength: Int) extends JavaAlgorithm {
+
+  import CryptoError.nonFatalHandling
+
+  val generateKeyPair: Crypto.KeyPairGenerator =
+    new Crypto.Func[Option[Array[Byte]], KeyPair] {
+      override def apply[F[_]](
+        input: Option[Array[Byte]]
```
Comment (alari) [inline hashsign/jvm/src/main/scala/fluence/crypto/ecdsa/Ed25519.scala:45]: If the input is provided, it should be used as a seed for the `SecureRandom`

File: hashsign/jvm/src/test/scala/fluence/crypto/Ed25519Spec.scala [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+import cats.instances.try_._
+import fluence.crypto.ecdsa.Ed25519
+import fluence.crypto.keystore.FileKeyStorage
+import fluence.crypto.signature.Signature
+import org.scalatest.{Matchers, WordSpec}
+import scodec.bits.ByteVector
+
+import scala.util.{Random, Try}
+
+class Ed25519Spec extends WordSpec with Matchers {
```
Comment (alari) [inline hashsign/jvm/src/test/scala/fluence/crypto/Ed25519Spec.scala:32]: Do we have any reference data & signature, e.g. from Tendermint? To check that everything matches.

**Review by alari — DISMISSED** [review]

File: hashsign/jvm/src/main/scala/fluence/crypto/ecdsa/Ed25519.scala [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+      )(implicit F: Monad[F]): EitherT[F, CryptoError, fluence.crypto.KeyPair] =
+        for {
+          g ← getKeyPairGenerator
+          random = input.map(new SecureRandom(_)).getOrElse(new SecureRandom())
+          keyParameters = new KeyGenerationParameters(random, strength)
+          _ = g.init(keyParameters)
+          p ← EitherT.fromOption(Option(g.generateKeyPair()), CryptoError("Could not generate KeyPair. Unexpected."))
```
Comment (folex) [inline hashsign/jvm/src/main/scala/fluence/crypto/ecdsa/Ed25519.scala]: improve error msg

File: hashsign/jvm/src/main/scala/fluence/crypto/ecdsa/Ed25519.scala [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+          keyPair ← nonFatalHandling {
+            val pk = p.getPublic.asInstanceOf[Ed25519PublicKeyParameters].getEncoded
+            val sk = p.getPrivate.asInstanceOf[Ed25519PrivateKeyParameters].getEncoded
+            KeyPair.fromBytes(pk, sk)
+          }("Could not generate KeyPair. Unexpected.")
```
Comment (folex) [inline hashsign/jvm/src/main/scala/fluence/crypto/ecdsa/Ed25519.scala]: error msg

File: hashsign/jvm/src/main/scala/fluence/crypto/ecdsa/Ed25519.scala [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+   * The public key will be the same each method call with the same secret key.
+   * @param sk secret key
+   * @return key pair
+   */
+  def restorePairFromSecret[F[_]: Monad](sk: Secret): EitherT[F, CryptoError, KeyPair] =
+    for {
+      keyPair ← nonFatalHandling {
+        val secret = new Ed25519PrivateKeyParameters(sk.bytes, 0)
+        KeyPair.fromBytes(secret.generatePublicKey().getEncoded, sk.bytes)
+      }("Could not generate KeyPair from private key. Unexpected.")
```
Comment (folex) [inline hashsign/jvm/src/main/scala/fluence/crypto/ecdsa/Ed25519.scala]: error msg

**Review by folex — APPROVED** [review]

**Pull request merged**